{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d660335c-6102-491b-9d2a-b633aa408d35",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "from plotly import graph_objects as go\n",
    "from pathlib import Path\n",
    "import os\n",
    "if not Path(\"jw_utils\").exists():\n",
    "    !git clone https://github.com/JonWinkelman/jw_utils\n",
    "from jw_utils import parse_fasta as pfa\n",
    "from jw_utils import parse_gff as pgf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2c4ab259-e312-4406-ba5a-5c798f310d0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "comp_genome_1 = Path(\"./data/3FF69J_results/3FF69J_1_1.1/3FF69J_1_1.1_1_reference.fasta\")\n",
    "comp_genome_2 = Path(\"./data/3FF69J_results/3FF69J_2_2.2/3FF69J_2_2.2_1_reference.fasta\")\n",
    "ref_genome    = Path(\"./references/GCA_000196175.1_ASM19617v1_genomic.fna\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "decde18f-e91d-40e2-964e-beebb73e0a33",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building alignments\n",
      "1: PREPARING DATA\n",
      "2,3: RUNNING mummer AND CREATING CLUSTERS\n",
      "# reading input file \"out.ntref\" of length 3782951\n",
      "# construct suffix tree for sequence of length 3782951\n",
      "# (maximum reference length is 2305843009213693948)\n",
      "# (maximum query length is 18446744073709551615)\n",
      "# process 37829 characters per dot\n",
      "#....................................................................................................\n",
      "# CONSTRUCTIONTIME /Users/jonathanwinkelman/miniconda3/envs/mummer/opt/mummer-3.23/mummer out.ntref 0.61\n",
      "# reading input file \"/Users/jonathanwinkelman/Trestle/Mukherjee_lab/20260105_analiml_bdello_genome_comp/mummer_genome_comparison/data/3FF69J_results/3FF69J_1_1.1/3FF69J_1_1.1_1_reference.fasta\" of length 3789023\n",
      "# matching query-file \"/Users/jonathanwinkelman/Trestle/Mukherjee_lab/20260105_analiml_bdello_genome_comp/mummer_genome_comparison/data/3FF69J_results/3FF69J_1_1.1/3FF69J_1_1.1_1_reference.fasta\"\n",
      "# against subject-file \"out.ntref\"\n",
      "# COMPLETETIME /Users/jonathanwinkelman/miniconda3/envs/mummer/opt/mummer-3.23/mummer out.ntref 2.31\n",
      "# SPACE /Users/jonathanwinkelman/miniconda3/envs/mummer/opt/mummer-3.23/mummer out.ntref 7.34\n",
      "4: FINISHING DATA\n",
      "Filtering alignments\n",
      "Extracting alignment coordinates\n",
      "Analyzing SNPs\n",
      "Extracting alignment breakpoints\n",
      "Generating report file\n",
      "gnuplot 5.4 patchlevel 10\n",
      "Writing filtered delta file out.filter\n",
      "Reading delta file out.delta\n",
      "Writing plot files out.fplot, out.rplot\n",
      "Writing gnuplot script out.gp\n",
      "Rendering plot out.png\n",
      "Building alignments\n",
      "1: PREPARING DATA\n",
      "2,3: RUNNING mummer AND CREATING CLUSTERS\n",
      "# reading input file \"out.ntref\" of length 3782951\n",
      "# construct suffix tree for sequence of length 3782951\n",
      "# (maximum reference length is 2305843009213693948)\n",
      "# (maximum query length is 18446744073709551615)\n",
      "# process 37829 characters per dot\n",
      "#....................................................................................................\n",
      "# CONSTRUCTIONTIME /Users/jonathanwinkelman/miniconda3/envs/mummer/opt/mummer-3.23/mummer out.ntref 0.61\n",
      "# reading input file \"/Users/jonathanwinkelman/Trestle/Mukherjee_lab/20260105_analiml_bdello_genome_comp/mummer_genome_comparison/data/3FF69J_results/3FF69J_2_2.2/3FF69J_2_2.2_1_reference.fasta\" of length 3775519\n",
      "# matching query-file \"/Users/jonathanwinkelman/Trestle/Mukherjee_lab/20260105_analiml_bdello_genome_comp/mummer_genome_comparison/data/3FF69J_results/3FF69J_2_2.2/3FF69J_2_2.2_1_reference.fasta\"\n",
      "# against subject-file \"out.ntref\"\n",
      "# COMPLETETIME /Users/jonathanwinkelman/miniconda3/envs/mummer/opt/mummer-3.23/mummer out.ntref 2.33\n",
      "# SPACE /Users/jonathanwinkelman/miniconda3/envs/mummer/opt/mummer-3.23/mummer out.ntref 7.33\n",
      "4: FINISHING DATA\n",
      "Filtering alignments\n",
      "Extracting alignment coordinates\n",
      "Analyzing SNPs\n",
      "Extracting alignment breakpoints\n",
      "Generating report file\n",
      "gnuplot 5.4 patchlevel 10\n",
      "Writing filtered delta file out.filter\n",
      "Reading delta file out.delta\n",
      "Writing plot files out.fplot, out.rplot\n",
      "Writing gnuplot script out.gp\n",
      "Rendering plot out.png\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def make_dnadiff_comp(ref_genome, comp_genome, outdir):\n",
    "    outdir.mkdir(exist_ok=True, parents=True)\n",
    "    comp = comp_genome.absolute()\n",
    "    ref = ref_genome.absolute()\n",
    "    working_dir = os.getcwd()\n",
    "    os.chdir(outdir)\n",
    "    !dnadiff $ref $comp\n",
    "    !mummerplot --png  --layout --fat out.delta\n",
    "    os.chdir(working_dir)\n",
    "\n",
    "\n",
    "outdir_1 = Path(f'./results/dnadiff/{comp_genome_1.stem}')\n",
    "make_dnadiff_comp(ref_genome, comp_genome_1, outdir_1)\n",
    "outdir_2 = Path(f'./results/dnadiff/{comp_genome_2.stem}')\n",
    "make_dnadiff_comp(ref_genome, comp_genome_2, outdir_2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc0d11d6-5a48-43e9-a90c-ecb655ed59df",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "source": [
    "| Col # | Meaning                           |\n",
    "| ----- | --------------------------------- |\n",
    "| 1     | `ref_pos` — position in reference |\n",
    "| 2     | `ref_base`                        |\n",
    "| 3     | `query_base`                      |\n",
    "| 4     | `query_pos`                       |\n",
    "| 5     | `ref_pos_in_aln`                  |\n",
    "| 6     | `query_pos_in_aln`                |\n",
    "| 7     | `ref_contig_len`                  |\n",
    "| 8     | `query_contig_len`                |\n",
    "| 9     | `ref_strand`                      |\n",
    "| 10    | `query_strand`                    |\n",
    "| 11    | `ref_contig`                      |\n",
    "| 12    | `query_contig`                    |  \n",
    "\n",
    ". in ref_base → insertion in query  \n",
    ". in query_base → deletion in query  \n",
    "query_strand = -1 → alignment is reversed (inversion or opposite orientation)  \n",
    "Multiple rows with same ref_pos → multi-base indels (expected)  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "31a79ce9-25cb-49f5-a5af-5ead2b2c2a02",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "def make_snp_df(snps_fp):\n",
    "    snps_df = pd.read_csv(\n",
    "        snps_fp,\n",
    "        sep=\"\\t\",\n",
    "        header=None\n",
    "    )\n",
    "    \n",
    "    snps_df.columns = [\n",
    "        \"ref_pos\",\n",
    "        \"ref_base\",\n",
    "        \"query_base\",\n",
    "        \"query_pos\",\n",
    "        \"ref_pos_aln\",\n",
    "        \"query_pos_aln\",\n",
    "        \"ref_contig_len\",\n",
    "        \"query_contig_len\",\n",
    "        \"ref_strand\",\n",
    "        \"query_strand\",\n",
    "        \"ref_contig\",\n",
    "        \"query_contig\",\n",
    "    ]\n",
    "    return snps_df\n",
    "snps_fp1 = outdir_1 / 'out.snps'\n",
    "snps_df1  = make_snp_df(snps_fp1)\n",
    "snps_df1.to_csv(outdir_1 /  'out.snps.csv')\n",
    "\n",
    "snps_fp2 = outdir_2 / 'out.snps'\n",
    "snps_df2  = make_snp_df(snps_fp2)\n",
    "snps_df2.to_csv(outdir_2 /  'out.snps.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e6637b5e-395b-464f-aa47-7041e477c515",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_genome_d = pfa.get_seq_dict(ref_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "3b80f1b8-c92d-461d-bbb4-d67663f934d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_ID</th>\n",
       "      <th>common_name</th>\n",
       "      <th>product</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>contig</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene-Bd0001</th>\n",
       "      <td>cds-CAE77685.1</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>chromosomal replication initiator protein dnaA</td>\n",
       "      <td>1</td>\n",
       "      <td>1416</td>\n",
       "      <td>+</td>\n",
       "      <td>BX842601.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-Bd0002</th>\n",
       "      <td>cds-CAE77686.1</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>DNA polymerase III%2C beta subunit</td>\n",
       "      <td>1649</td>\n",
       "      <td>2755</td>\n",
       "      <td>+</td>\n",
       "      <td>BX842601.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-Bd0003</th>\n",
       "      <td>cds-CAE77687.1</td>\n",
       "      <td>Bd0003</td>\n",
       "      <td>DNA repair and genetic recombination protein</td>\n",
       "      <td>2755</td>\n",
       "      <td>3879</td>\n",
       "      <td>+</td>\n",
       "      <td>BX842601.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-Bd0004</th>\n",
       "      <td>cds-CAE77688.1</td>\n",
       "      <td>Bd0004</td>\n",
       "      <td>DNA gyrase subunit B</td>\n",
       "      <td>3966</td>\n",
       "      <td>6386</td>\n",
       "      <td>+</td>\n",
       "      <td>BX842601.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene-Bd0005</th>\n",
       "      <td>cds-CAE77689.1</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>DNA gyrase subunit A</td>\n",
       "      <td>6410</td>\n",
       "      <td>8893</td>\n",
       "      <td>+</td>\n",
       "      <td>BX842601.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 protein_ID common_name  \\\n",
       "gene_ID                                   \n",
       "gene-Bd0001  cds-CAE77685.1        dnaA   \n",
       "gene-Bd0002  cds-CAE77686.1        dnaN   \n",
       "gene-Bd0003  cds-CAE77687.1      Bd0003   \n",
       "gene-Bd0004  cds-CAE77688.1      Bd0004   \n",
       "gene-Bd0005  cds-CAE77689.1        gyrA   \n",
       "\n",
       "                                                    product  start   end  \\\n",
       "gene_ID                                                                    \n",
       "gene-Bd0001  chromosomal replication initiator protein dnaA      1  1416   \n",
       "gene-Bd0002              DNA polymerase III%2C beta subunit   1649  2755   \n",
       "gene-Bd0003    DNA repair and genetic recombination protein   2755  3879   \n",
       "gene-Bd0004                            DNA gyrase subunit B   3966  6386   \n",
       "gene-Bd0005                            DNA gyrase subunit A   6410  8893   \n",
       "\n",
       "            strand      contig  \n",
       "gene_ID                         \n",
       "gene-Bd0001      +  BX842601.2  \n",
       "gene-Bd0002      +  BX842601.2  \n",
       "gene-Bd0003      +  BX842601.2  \n",
       "gene-Bd0004      +  BX842601.2  \n",
       "gene-Bd0005      +  BX842601.2  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_to_gff = Path('./references/genomic.gff')\n",
    "annot_df = pgf.make_simple_annot_df(path_to_gff, start_end=True, contig=True)\n",
    "annot_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "6f202b08-5219-48f4-9055-763493d3d6ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bisect\n",
    "def map_snps_to_genes(snps_df, annot_df): \n",
    "    \n",
    "    colnames  = [\n",
    "        \"protein_ID\",\n",
    "        \"common_name\",\n",
    "        \"product\",\n",
    "        \"start\",\n",
    "        \"end\",\n",
    "        \"strand\",\n",
    "        \"contig\",\n",
    "        'location', \n",
    "        'snp_id',\n",
    "        'snp_position'\n",
    "        ]\n",
    "    cols_d = {c:[] for c in colnames}\n",
    "    starts = list(annot_df['start'])\n",
    "    ends = list(annot_df['end'])\n",
    "    \n",
    "    \n",
    "    \n",
    "    for ind, row in snps_df.iterrows():\n",
    "        start_index = bisect.bisect_left(starts, row['ref_pos'])\n",
    "        end_index   = bisect.bisect_left(ends, row['ref_pos'])\n",
    "        s = annot_df.iloc[end_index]\n",
    "        cols_d['snp_id'].append(ind)\n",
    "        cols_d['snp_position'].append(row['ref_pos'])\n",
    "        for col in colnames:\n",
    "            if col in s.index:\n",
    "                cols_d[col].append(s[col])\n",
    "            elif col == 'location':\n",
    "                if start_index-1 == end_index: \n",
    "                    location = 'within_gene'\n",
    "                    cols_d[col].append(location)\n",
    "                else:\n",
    "                    location = 'intergenic'\n",
    "                    cols_d[col].append(location)\n",
    "    return snps_df.join(pd.DataFrame(cols_d).set_index('snp_id'))           \n",
    "\n",
    "snps_annot_df1 = map_snps_to_genes(snps_df1, annot_df)\n",
    "snps_annot_df2 = map_snps_to_genes(snps_df2, annot_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "f6be6b6b-3fdf-4c46-9760-099c0bf5a909",
   "metadata": {},
   "outputs": [],
   "source": [
    "snps_annot_df1.to_csv(outdir_1 / 'out.snps_annot.csv')\n",
    "snps_annot_df2.to_csv(outdir_2 / 'out.snps_annot.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d1ad0a6-bd4c-4e7b-841e-9a86d202000e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2125668e-e0e1-418a-a6ad-685bf23749a5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3168bc1e-4edd-453b-8d8f-7733bfaa7efd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
